lib/pos_mempool.go

```diff
@@ -0,0 +1,490 @@
+	}
+	// Reset the transaction register, the ledger, and the nonce tracker.
+	dmp.txnRegister.Reset()
+	dmp.ledger.Reset()
+	dmp.nonceTracker.Reset()
+
+	dmp.status = PosMempoolStatusNotRunning
+}
+
+func (dmp *PosMempool) IsRunning() bool {
+	return dmp.status == PosMempoolStatusRunning
+}
+
+// AddTransaction validates a MsgDeSoTxn transaction and adds it to the mempool if it is valid.
+// If the mempool overflows as a result of adding the transaction, the mempool is pruned.
+func (dmp *PosMempool) AddTransaction(txn *MsgDeSoTxn) error {
+	// First, validate that the transaction is properly formatted according to BalanceModel. We acquire a read lock on
+	// the mempool. This allows multiple goroutines to safely perform transaction validation concurrently. In particular,
+	// transaction signature verification can be parallelized.
+	if err := dmp.validateTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem verifying transaction")
+	}
+
+	// If we get this far, it means that the transaction is valid. We can now add it to the mempool.
+	// We lock the mempool to ensure that no other thread is modifying it while we add the transaction.
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.AddTransaction: ")
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, dmp.latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check the nonceTracker to see if this transaction is meant to replace an existing one.
+	pk := NewPublicKey(txn.PublicKey)
+	if existingTxn := dmp.nonceTracker.GetTxnByPublicKeyNonce(*pk, *txn.TxnNonce); existingTxn != nil {
+		if existingTxn.FeePerKB > mempoolTx.FeePerKB {
+			return errors.Wrapf(MempoolFailedReplaceByHigherFee, "PosMempool.AddTransaction: Problem replacing transaction "+
+				"by higher fee failed. New transaction has lower fee.")
+		}
+		if err := dmp.removeTransactionNoLock(existingTxn, true); err != nil {
+			return errors.Wrapf(err, "PosMempool.AddTransaction: Problem removing old transaction from mempool during "+
+				"replacement with higher fee.")
+		}
+	}
+
+	// Add the transaction to the mempool and then prune if needed.
+	if err := dmp.addTransactionNoLock(mempoolTx, true); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	// Add the transaction to the nonce tracker.
+	dmp.nonceTracker.AddTxnByPublicKeyNonce(*pk, *txn.TxnNonce, mempoolTx)
+
+	if err := dmp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) validateTransaction(txn *MsgDeSoTxn) error {
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	if err := CheckTransactionSanity(txn, uint32(dmp.latestBlockHeight), dmp.params); err != nil {
```

Comment:
I'm not sure CheckTransactonSanity is doing anything essential anymore. Maybe we can delete it?

Reply:
I'm a little worried to take it out. It's doing some sanity-checking to transaction inputs that seem not entirely obsolete. I think it's a "safe bet" to keep it.